eyieko/landville-frontend
# #168164365 setup a service worker

## Description ##


This Pull Request helps improve the application's performance by integrating a service worker to help cache static assets such as images, HTML , CSS and JS files.


## Type of change ##
Please select the relevant option

## How Has This Been Tested? ##
Please describe the tests that you ran to verify your changes

## Checklist: ##

## How can this PR be manually tested: ##

### To test this on the PR app 
1. Open [Link](https://landville-frontend-pr-46.herokuapp.com/home)
2. You will see a plus icon in the `omni-box` upon clicking it, it will ask you if you want to install the `landville app`. Note that PWA's can only be installed if the browser ascertains that a service worker is present and a manifest.json is also present in the app.
![image](https://user-images.githubusercontent.com/12128153/64155692-f75d7400-ce3b-11e9-90ad-a86a6144fcc3.png)

### To test this locally
1. Clone this repo and run `npm install`
2. Run `ng build --prod` or `npm run heroku-postbuild` 
3. Run `npm start`
4. Open the browser on `http://localhost:8080/home`
5. Reload the page 
6. Open up the devtools and head on over to the application tab and on the service worker section, you will see a service worker has been activated

![image](https://user-images.githubusercontent.com/12128153/64029988-5c6c4d80-cb4e-11e9-8529-3c29df895261.png)

7. Visit a couple of properties of your choosing and head on back to the `devtools` and in the `application tab` open up the `cache storage` open up the different `cache entries` in the list and you will find `images`, `html`, `css` and `js` files have been stored

![image](https://user-images.githubusercontent.com/12128153/64030163-b5d47c80-cb4e-11e9-8483-b38d46e44632.png)

## Any background context you want to provide? ##

Service workers enable applications to control network requests, cache those requests to improve performance and provide offline access to cached content. It's important to be considerate of the users of our application, some may not have strong internet connections and service workers may help in making the app fast and responsive for them by caching assets and retrieving the assets from the cache on repeat visits instead of making new network requests 

### Performance measurement:
#### With service workers
With service workers, a return visit to the app sees it load in under `a second`

![image](https://user-images.githubusercontent.com/12128153/64096581-bb0e1300-cd6a-11e9-81ed-0e34767a4d57.png)
#### Without service workers

On Return/repeat visits the app loads 4 to 5 times slower

![image](https://user-images.githubusercontent.com/12128153/64096740-193af600-cd6b-11e9-955e-39af6bdf90b7.png)

The chrome extension used to measure page speeds is [Page load time](https://chrome.google.com/webstore/detail/page-load-time/fploionmjgeclbkemipmkogoaohcdbig?hl=en)




## Related Pivotal Tracker stories ##
[#168164365](https://www.pivotaltracker.com/n/projects/2343772/stories/168164365)
